discojs/src/privacy.ts

```diff
@@ -1,46 +1,39 @@
-import * as tf from '@tensorflow/tfjs'
+import * as tf from "@tensorflow/tfjs";
 
-import type { Task, WeightsContainer } from './index.js'
+import type { WeightsContainer } from "./index.js";
 
-/**
- * Add task-parametrized Gaussian noise to and clip the weights update between the previous and current rounds.
- * The previous round's weights are the last weights pulled from server/peers.
- * The current round's weights are obtained after a single round of training, from the previous round's weights.
- * @param updatedWeights weights from the current round
- * @param staleWeights weights from the previous round
- * @param task the task
- * @returns the noised weights for the current round
- */
-export function addDifferentialPrivacy (updatedWeights: WeightsContainer, staleWeights: WeightsContainer, task: Task): WeightsContainer {
-  const noiseScale = task.trainingInformation?.noiseScale
-  const clippingRadius = task.trainingInformation?.clippingRadius
+async function frobeniusNorm(weights: WeightsContainer): Promise<number> {
```

Comment:
isn't this one a duplicate of norm in the spec one now?

Reply:
```suggestion
// The Frobenius norm of a matrix is the square root of the sum of squared weights
async function frobeniusNorm(weights: WeightsContainer): Promise<number> {
```
Also tfjs has `tf.norm(x, 'fro')` which already does frobenius norm, though this would not be backend-agnostic. https://js.tensorflow.org/api/latest/#norm

Reply:
tf.norm could be a way to test the hand-made implementation

Reply:
yep indeed. my personnal duplication-limit is at 3 for small functions, I'm extract it into a common one if we need it somewhere else.